### PR TITLE
feat: add option for automatic account linking

### PR DESCRIPTION
**Pull request by hperl**
## Checklist

- [ ] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [ ] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [ ] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [ ] I have read the [security policy](../security/policy).
- [ ] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got the approval (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [ ] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](https://github.com/ory/docs).

### CODE REVIEW

File: selfservice/strategy/oidc/strategy_registration.go [hunk 1]
```diff
@@ -349,13 +362,57 @@ func (s *Strategy) processRegistration(w http.ResponseWriter, r *http.Request, r
...
 	if err := s.d.RegistrationExecutor().PostRegistrationHook(w, r, identity.CredentialsTypeOIDC, provider.Config().ID, rf, i); err != nil {
+		dup := new(identity.ErrDuplicateCredentials)
+		if provider.Config().AutomaticAccountLinking == AutomaticAccountLinkingReplace && errors.As(err, &dup) {
+			conflictingIdentity, _, err := s.d.IdentityManager().FindConflictingIdentity(r.Context(), i)
+			if err != nil {
+				return nil, s.handleError(w, r, rf, provider.Config().ID, i.Traits, err)
+			}
+			if conflictingIdentity == nil {
+				return nil, s.handleError(w, r, rf, provider.Config().ID, i.Traits, dup)
+			}
```
Comment (jonas-jonas) [inline selfservice/strategy/oidc/strategy_registration.go:373]: Can this actually ever happen? Returning `nil, "", nil` from `FindConflictingIdentity` can lead to panics if you forget to check the value of the identity for nil, which you don't typically do. E.g. you assume that the values returned are valid if no `err` was returned. Maybe we can return an error from `FindConflictingIdentity` instead?

Comment (hperl) [reply selfservice/strategy/oidc/strategy_registration.go:373]: Yeah that sounds sensible. I just extracted the method and the `nil, "", nil` case was when we could not find a conflicting identity. I guess an error is better in this case.

**Review by aeneasr — COMMENTED** [review]
Can you please create an E2E test for this and briefly explain what the user will see?

File: identity/identity.go [hunk 2]
```diff
@@ -160,6 +160,24 @@ func (t *Traits) String() string {
 	return string(*t)
 }
 
+// Merge merges the other traits into this one. If a key exists in both, the
+// other traits' value is use
+func (t *Traits) Merge(o Traits) (Traits, error) {
```
Comment (aeneasr) [inline identity/identity.go:165]: This only merges the first level. Is that what we want? If not, this should be something like mapstructure.Merge or similar.

File: identity/manager.go [hunk 3]
```diff
@@ -110,12 +110,11 @@ func (m *Manager) Create(ctx context.Context, i *Identity, opts ...ManagerOption
 	return nil
 }
 
-func (m *Manager) findExistingAuthMethod(ctx context.Context, e error, i *Identity) (err error) {
-	if !m.r.Config().SelfServiceFlowRegistrationLoginHints(ctx) {
-		return &ErrDuplicateCredentials{error: e}
-	}
+func (m *Manager) FindConflictingIdentity(ctx context.Context, i *Identity) (found *Identity, foundConflictAddress string, err error) {
```
Comment (aeneasr) [inline identity/manager.go:113]: Test pls

File: identity/identity.go (hunk 2 above)
Comment (aeneasr) [reply identity/identity.go:165]: Also test please

File: embedx/config.schema.json [hunk 4]
```diff
@@ -603,6 +603,16 @@
           "examples": [
             "12345678-1234-1234-1234-123456789012"
           ]
+        },
+        "automatic_account_linking": {
+          "title": "Configure automatic account linking",
+          "description": "Configures how accounts through this provider should be linked automatically to existing accounts.",
```
Comment (aeneasr) [inline embedx/config.schema.json:609]: What happens when this is "never" and what if it is "replace"?

File: selfservice/strategy/oidc/provider_config.go [hunk 5]
```diff
@@ -108,8 +108,15 @@ type Configuration struct {
 	// An optional organization ID that this provider belongs to.
 	// This parameter is only effective in the Ory Network.
 	OrganizationID string `json:"organization_id"`
+
+	AutomaticAccountLinking string `json:"automatic_account_linking"`
```
Comment (aeneasr) [inline selfservice/strategy/oidc/provider_config.go:112]: Docstring please

And potentially use a dedicated type if this is an enum

File: selfservice/strategy/oidc/strategy_registration.go [hunk 6]
```diff
@@ -349,13 +362,62 @@ func (s *Strategy) processRegistration(w http.ResponseWriter, r *http.Request, r
 
 	i.SetCredentials(s.ID(), *creds)
 	if err := s.d.RegistrationExecutor().PostRegistrationHook(w, r, identity.CredentialsTypeOIDC, provider.Config().ID, rf, i); err != nil {
+		dup := new(identity.ErrDuplicateCredentials)
+		if provider.Config().AutomaticAccountLinking == AutomaticAccountLinkingReplace && errors.As(err, &dup) {
```
Comment (aeneasr) [inline selfservice/strategy/oidc/strategy_registration.go:366]: fyi, you can also do:

```
		if dup := new(identity.ErrDuplicateCredentials); provider.Config().AutomaticAccountLinking == AutomaticAccountLinkingReplace && errors.As(err, &dup) {
```

File: selfservice/strategy/oidc/strategy_registration.go (hunk 1 above)
Comment (aeneasr) [reply selfservice/strategy/oidc/strategy_registration.go:373]: Agree, `nil, "", nil` should not be a valid return type of FindConflictingIdentity

**Review by aeneasr — CHANGES_REQUESTED** [review]
Thank you!

I believe that this PR falls into the trap why we do not have automatic account linking in the first place, which is documented here: https://www.ory.sh/docs/kratos/social-signin/link-multiple-provider-account#automatic-account-linking

Please adjust the PR so that this can not happen, or comment on why the implementation is not an exploitable attack vector. Thanks!

File: embedx/config.schema.json [hunk 7]
```diff
@@ -603,6 +603,16 @@
...
+          "type": "string",
+          "enum": [
+            "never",
+            "replace"
```
Comment (alnr) [inline embedx/config.schema.json:613]:
```suggestion
            "replace_all_other_credentials"
```

**Comment by aeneasr** [comment]
The fundamental problem of this approach is that we (Ory Kratos) does not know whether identity A and identity B are truly the same person or not. Just because some credential is saying that they are the same user, does not mean that they indeed are the same person. In the end, this is an account takeover vulnerability which essentially:

1. Alice signs up with email and password `alice@doe.com`
2. Eve does account enumeration (currently possible) or otherwise knows that alice has an account
3. Eve signs up at the social provider that is enabled with account linking (e.g. GitHub) with `alice@doe.com`
4. Eve logs into Ory Kratos using GitHub with the linked fake `alice@doe.com` email
5. The original password is deleted, and Eve has a valid session token -> account takeover complete

There are probably some ways of fixing this with verified email addresses and so on. But honestly, let's just stick to how we solve it right now but make it more user friendly:

1. Sign up with email password
2. Sign up with social provider -> conflict
3. Present user login screen "sign in with your password to link your account"
4. Sign in with correct credential is complete
5. Social sign in provider is linked to the user

**Comment by kmherrmann** [comment]
> The fundamental problem of this approach is that we (Ory Kratos) does not know whether identity A and identity B are truly the same person or not. Just because some credential is saying that they are the same user, does not mean that they indeed are the same person. In the end, this is an account takeover vulnerability which essentially:
> 
> 1. Alice signs up with email and password `alice@doe.com`
> 2. Eve does account enumeration (currently possible) or otherwise knows that alice has an account
> 3. Eve signs up at the social provider that is enabled with account linking (e.g. GitHub) with `alice@doe.com`
> 4. Eve logs into Ory Kratos using GitHub with the linked fake `alice@doe.com` email
> 5. The original password is deleted, and Eve has a valid session token -> account takeover complete
> 
> There are probably some ways of fixing this with verified email addresses and so on. But honestly, let's just stick to how we solve it right now but make it more user friendly:

Yes, we can only do this automatically if we trust that we have verified email addresses on both sides. For that, I think we should have a flag on OIDC providers `like trust_email_verification`


> 
> 1. Sign up with email password
> 2. Sign up with social provider -> conflict
> 3. Present user login screen "sign in with your password to link your account"
> 4. Sign in with correct credential is complete
> 5. Social sign in provider is linked to the user

I think this flow is great and should be the default.

In the case that the existing Ory account has a verified email address AND the OIDC login claims to have a verified email address AND the OIDC provider is set to trust email verification, then the steps 3 and 4 can automatically be skipped.

**Pull request closed without merging**